### ISSUE TITLE
Unexpected error when spreading object with optional properties

### ISSUE THREAD

**Issue by OliverJAsh** [Microsoft/TypeScript#16445]
**TypeScript Version:**  2.3.4

**Code**

```ts
{
    type MaybeFoo = { foo?: string };
    type Foo = { foo: string };

    const z: MaybeFoo = {};
    const y: Foo = { foo: 'bar' };
    const x: Foo = { ...y, ...z };
    // Error:
    // Type '{ foo: string | undefined; }' is not assignable to type 'Foo'.
    //   Types of property 'foo' are incompatible.
    //     Type 'string | undefined' is not assignable to type 'string'.
    //       Type 'undefined' is not assignable to type 'string'.
}
```

**Expected behavior:**
I did not expect this error, because `MaybeFoo` marks the `foo` property as optional, not as `string | undefined`.

For context, I want this so I can define optional and non-optional properties for an object, with defaults for the optional ones, before merging the defaults with the provided properties:

``` ts
{
    type All = { foo: string, bar: string };
    type Requireds = Pick<All, 'foo'>;
    type Defaults = Pick<All, 'bar'>;
    type Args = Requireds & Partial<Defaults>;
    const defaults: Defaults = { bar: 'bar' };

    const fn = (args: Args) => {
        const all: All = { ...defaults, ...args }; // Unexpected error
    };
}
```

---

I'm also interested to know why this does not error:

``` ts
{
    type MaybeFoo = {
        foo?: string,
    };

    const x: MaybeFoo = {
        foo: undefined,
    };
}
```

I thought optional properties were distinct from nullable types, because they indicate a missing key, whereas the latter indicates a defined key with a nullable value.

**Comment by johnsoft** [comment]
I ran into this trying to upgrade from 2.2.2 to 2.4.0. It's also present in `2.4.0-dev.20170612`. This breaks a fundamental use case of object spread in functional code:

```ts
interface Filter { color: string; region: string; /* etc */ }

function updateFilter(f1: Filter, f2: Partial<Filter>): Filter {
  return { ...f1, ...f2 };
}
```

**tsconfig.json**

`{"compilerOptions": {"strictNullChecks": true}}`

**Expected output:**

No errors

**Actual output:**

```
index.tsx(4,3): error TS2322: Type '{ color: string | undefined; region: string | undefined; }' is not assignable to type 'Filter'.
  Types of property 'color' are incompatible.
    Type 'string | undefined' is not assignable to type 'string'.
      Type 'undefined' is not assignable to type 'string'.
```

**Comment by mhegazy** [comment]
Duplicate of https://github.com/Microsoft/TypeScript/issues/16509

**Comment by OliverJAsh** [comment]
Now fixed, see #16509
